pkg/util/containers/runtime.go

```diff
@@ -0,0 +1,94 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// This code is not tied to docker itself, hence no docker build flag.
+// It could be moved to its own package.
+
+package containers
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/shirou/gopsutil/process"
+)
+
+// Return values per container runtime
+const (
+	RuntimeNameDocker     string = "docker"
+	RuntimeNameContainerd string = "containerd"
+	RuntimeNameCRIO       string = "cri-o"
+)
+
+// Internal constants
+const (
+	daemonNameDockerLegacy1  string = "dockerd"
+	daemonNameDockerLegacy2  string = "dockerd-current" // CentOS
+	shimNameContainerd       string = "containerd-shim"
+	shimNameCRIO             string = "conmon"
+	shimNameContainerdUnsure string = "docker-containerd-shim"
+	shimArgContainerdK8s     string = "-namespace k8s.io"
+	shimArgContainerdDocker  string = "-namespace moby"
+)
+
+// ErrNoRuntimeMatch is returned when no container runtime can be matched
+var ErrNoRuntimeMatch = errors.New("cannot match a container runtime")
+
+// GetRuntimeForPID inspects a PID's parents to detect a container runtime.
+// For now, this assumes we are running on hostPID, as gopsutil looks-up
+// processess in `/proc` (or HOST_PROC if set)
+func GetRuntimeForPID(pid int32) (string, error) {
+	var currentProcess *process.Process
+	// Inspect given process
+	currentProcess, err := process.NewProcess(pid)
+	if err != nil {
+		return "", err
+	}
+
+	for {
+		// Get process cmdline and extract cmd name. Not using the `exe`
+		// symlink because we don't always have permissions to read it
+		cmdline, err := currentProcess.CmdlineSlice()
+		if err != nil {
+			return "", err
+		}
+		cmd := cmdline[0]
```

Comment:
Is there some edge-case where the this would fail? like `nohup`? I don't know if it's legit but worth thinking about

Reply:
Actually `nohup` or `exec` don't show up in the process command line tree so we should be fine, I don't see use-cases where `cmdline[0]` would not be what we're looking for

Reply:
These process wrappers either:
  - `fork` their invocation as a new process (example: tini), in that case we'll get the good `cmdline` on that process
  - `exec` into the process, changing their `cmdline` to the actual invocation

So we should be safe